File: Python/34.Search for a Range.py
#给出数字数组中，目标数字第一个和最后一个出现的位置，没有则返回[-1,1]
#要求算法O(log n)
class Solution(object):
    def searchRange(self, nums, target):
        l , r = 0 , len(nums)-1
        while l<=r:
            mid = (l+r)//2
            if nums[mid] == target:
                break
            elif nums[mid]<target:
                l = mid+1
            else:
                r = mid-1

        if nums[mid]==target:
            l = mid
            while l>=0 and nums[l]==nums[mid]:
                l-=1
            r = mid
            while r<len(nums) and nums[r]==nums[mid]:
                r+=1
            return [l+1,r-1]
        else:
            return [-1,-1]



s = Solution()
print(s.searchRange([5,7,7,8,8,10],10))

File: C++/Array/34.Search for a Range.cpp
/*
一个有序数组，里面有重复元素。
返回给定数在数组中的位置范围。如果不在，[-1,-1]。
要求时间O(logN)。
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
     vector<int> searchRange(vector<int>& nums, int target) {
     	vector<int> result{-1,-1};
     	if(nums.size()==0)	return result;
     	int left = 0,right = nums.size()-1,mid;
     	while(left<right)
     	{
     		mid = (left+right)/2;
     		if(nums[mid]==target)
     		{
     			left = mid;		break;
     		}
     		if(nums[mid]>target)	right = mid-1;
     		else	left = mid+1;
     	}
     	if(nums[left]!=target)	return result;
     	int i = left;
     	while(i>=0 && target==nums[i])	--i;
     	result[0] = i+1,i = left;
     	while(i<nums.size() && target==nums[i])	 ++i;
     	result[1] = i-1;
     	return result;
     }
};

int main()
{
	Solution s;
	vector<int> nums{5, 7, 7, 8, 8, 10};
	nums = s.searchRange(nums, 3);
	cout<<nums[0]<<","<<nums[1]<<endl;
}

